cmd/ephemeral-installer-register-cluster-and-infraenv/main.go

```diff
@@ -0,0 +1,248 @@
+	log.Info("releaseImage: " + getReleaseVersion())
+	log.Infof("releaseImage version %s cpuarch %s", releaseImageVersion, releaseImageCPUArch)
+
+	var infraEnv aiv1beta1.InfraEnv
+	if infraenvErr := getFileData(Options.InfraEnvFile, &infraEnv); infraenvErr != nil {
+		log.Fatal(infraenvErr.Error())
+	}
+
+	var nmStateConfig aiv1beta1.NMStateConfig
+	if nmStateErr := getFileData(Options.NMStateConfigFile, &nmStateConfig); nmStateErr != nil {
+		log.Fatal(nmStateErr.Error())
+	}
+
+	staticNetworkConfig, processErr := processNMStateConfig(log, infraEnv, nmStateConfig)
+	if processErr != nil {
+		log.Fatal(processErr.Error())
+	}
+
+	clientConfig := client.Config{}
+	u, parseErr := url.Parse(Options.ServiceBaseUrl)
+	if parseErr != nil {
+		log.Fatal(parseErr, "Failed parsing inventory URL")
+	}
+	u.Path = path.Join(u.Path, client.DefaultBasePath)
+	clientConfig.URL = u
+	bmInventory := client.New(clientConfig)
+	ctx := context.Background()
+
+	log.Info("SERVICE_BASE_URL: " + Options.ServiceBaseUrl)
+	log.Info("Registering cluster")
+
+	clusterParams := controllers.CreateClusterParams(&cd, &aci, pullSecret, releaseImageVersion, releaseImageCPUArch, nil)
+	clientClusterParams := &installer.V2RegisterClusterParams{
+		NewClusterParams: clusterParams,
+	}
+	clusterResult, registerClusterErr := bmInventory.Installer.V2RegisterCluster(ctx, clientClusterParams)
+	if registerClusterErr != nil {
+		log.Fatal(registerClusterErr, "Failed to register cluster with assisted-service")
+	}
+
+	modelsCluster := clusterResult.Payload
+
+	log.Info("Registered cluster with id: " + modelsCluster.ID.String())
+
+	log.Info("Registering infraenv")
+
+	infraEnvParams := controllers.CreateInfraEnvParams(&infraEnv, models.ImageType(Options.ImageTypeISO), pullSecret, modelsCluster.ID, modelsCluster.OpenshiftVersion)
+
+	if len(staticNetworkConfig) > 0 {
+		log.Infof("Added %d nmstateconfigs", len(staticNetworkConfig))
+		infraEnvParams.InfraenvCreateParams.StaticNetworkConfig = staticNetworkConfig
+	}
+
+	clientInfraEnvParams := &installer.RegisterInfraEnvParams{
+		InfraenvCreateParams: infraEnvParams.InfraenvCreateParams,
+	}
+	infraEnvResult, registerInfraEnvErr := bmInventory.Installer.RegisterInfraEnv(ctx, clientInfraEnvParams)
+	if registerInfraEnvErr != nil {
+		log.Fatal(registerInfraEnvErr, "Failed to register infraenv with assisted-service")
+	}
+
+	modelsInfraEnv := infraEnvResult.Payload
+
+	log.Info("Registered infraenv with id: " + modelsInfraEnv.ID.String())
+}
+
+func getReleaseVersion() string {
+	var clusterImageSet hivev1.ClusterImageSet
+	if err := getFileData(Options.ClusterImageSetFile, &clusterImageSet); err != nil {
+		log.Fatal(err.Error())
+	}
+	return clusterImageSet.Spec.ReleaseImage
+}
+
+func getReleaseVersionAndCpuArch(releaseImage string) (version string, cpuArch string) {
+	// releaseImage is in the form: quay.io/openshift-release-dev/ocp-release:4.9.17-x86_64
+	versionCpuArch := strings.Split(releaseImage, ":")[1]
```

Comment:
This can break when using port numbers, e.g:
`quay.io:443/openshift-release-dev/ocp-release:4.9.17-x86_64`

Reply:
Or when using exact sha:
`quay.io/openshift-release-dev/ocp-release@sha256:7b67b0cb5ab016528b8efdb6130c000398efc58f55e2226f3cf4e3be59c78443`

But admittedly the latter one is less of a problem because you wouldn't be able to parse the architecture anyway

Reply:
Good catch.